README.md
# App::CLI - Dispatcher module for command line interface programs

[![Coverage Status](https://coveralls.io/repos/github/paultcochrane/app-cli/badge.svg)
](https://coveralls.io/github/paultcochrane/app-cli)
[![Travis Build Status](https://travis-ci.org/paultcochrane/app-cli.svg?branch=master)
](https://travis-ci.org/paultcochrane/app-cli)
[![AppVeyor Build status](https://ci.appveyor.com/api/projects/status/g0dlp43k42lpy3s8?svg=true)
](https://ci.appveyor.com/project/paultcochrane/app-cli)


`App::CLI` dispatches CLI (command line interface) based commands
into command classes.  It also supports subcommand and per-command
options.

## Installation

The easiest way to install `App::CLI` is to use `cpanm`:

```
$ cpanm App::CLI
```

To install it from source, clone the repo, create the `Makefile`, and run
`make install`:

```
git clone https://github.com/paultcochrane/app-cli.git
cd app-cli
perl Makefile.PL
make
make test
make install
```

## Example usage

```
package MyApp;
use base 'App::CLI';        # the DISPATCHER of your App
                            # it's not necessary to put the dispatcher
                            # on the top level of your App

package main;

MyApp->dispatch;            # call the dispatcher where you want


package MyApp::List;
use base qw(App::CLI::Command); # any (SUB)COMMAND of your App

use constant options => (
    "h|help"   => "help",
    "verbose"  => "verbose",
    'n|name=s'  => 'name',
);

use constant subcommands => qw(User Nickname type); # if you want subcommands
                                                    # automatically dispatch to subcommands
                                                    # when invoke $ myapp list [user|nickname|--type]
                                                    # note 'type' is not capitalized
                                                    # it is a deprecated subcommand

sub run {
    my ($self, @args) = @_;

    print "verbose" if $self->{verbose};
    my $name = $self->{name}; # get arg following long option --name

    if ($self->{help}) {
        # if $ myapp list --help or $ myapp list -h
        # only output PODs
    } else {
        # do something when invoking $ myapp list
        # without subcommand and --help
    }
}


package MyApp::List::User;
use base qw(App::CLI::Command);
use constant options => (
    "h|help"  =>  "help",
);

sub run {
    my ($self,@args) = @_;
    # code for listing user
}


pakcage MyApp::List::Nickname;
use base qw(App::CLI::Command);
use constant options => (
    "sort=s"  =>  "sort",
);

sub run {
    my ($self,@args) = @_;
    # code for listing nickname
}

package MyApp::List::type;   # old genre of subcommand could not cascade infinitely
use base qw(MyApp::List);    # should inherit its parent's command

sub run {
    my ($self, @args);
    # run to here when invoking $ myapp list --type
}


package MyApp::Help;
use base 'App::CLI::Command::Help';

use constant options => (
    'verbose' => 'verbose',
);

sub run {
    my ($self, @arg) = @_;
    # do something
    $self->SUPER(@_); # App::CLI::Command::Help would output POD of each command
}
```


Changes
0.52 Tue Dec 15 11:27:33 CET 2020

  * set bugtracker metadata explicitly

0.51 Mon Dec 14 19:56:11 CET 2020

  * fix licensing inconsistency (spotted by kentnl)

0.50 Mon May 14 19:41:04 CEST 2018

  * improve 'brief_usage()' output (patch from Ildar Shaimordanov)

0.49 Mon Nov 06 07:46:11 CET 2017

  * add files to MANIFEST missing from last release

0.48 Sun Nov 05 13:34:46 CET 2017

  * fixed help output behaviour (patch from sunnavy)
  * more rigorous testing of Help class
  * minor documentation improvements
  * add AppVeyor configuration

0.47 Wed Nov 01 22:31:41 CET 2017

  * add missing prereqs
  * add missing LICENSE file
  * add POD tests
  * fix utf8 error in POD

0.46  Wed Nov 01 20:21:58 CEST 2017

  * many documentation updates
  * extended test suite
  * fix several perlcritic issues
  * mention package aliases in command list (José Joaquín Atria)
  * add default version command (José Joaquín Atria)
  * add a 'commands' command to list commands in app (José Joaquín Atria)
  * multiple general improvements to the code (José Joaquín Atria)

0.45  Wed Jun 28 23:05:58 CEST 2017

  * use a version number in the original style

0.4.5  Wed Jun 28 22:59:33 CEST 2017

  * put VERSION back in main module file

0.4.4  Wed 28 Jun 2017 22:52:33 CEST

  * remove 'provides' metadata

0.4.3  Wed 28 Jun 2017 22:37:58 CEST

  * put VERSION back into only Makefile.PL
  * updated author information

0.4.2  Wed 28 Jun 2017 22:25:29 CEST

  * add VERSION to all .pm files

0.4.1  Wed 28 Jun 2017 21:44:06 CEST

  * move version number into Makefile.PL

0.4  Wed 28 Jun 2017 21:37:23 CEST

  * version bump; PAUSE thinks that 0.313 is higher than 0.32

0.32 Wed 28 Jun 2017 20:30:19 CEST

  * Fix RT#122101: unskip META.yml
  * Fix RT#120982: can't find inc/Module/Install.pm; replaced
    Module::Install with ExtUtils::MakeMaker
  * Fix RT#63798: fix typo in manpage
  * Fix RT#62241: signature broken (signature has been removed)
  * Fixed more typos in POD
  * Fixed some PerlCritic issues
  * Updated dependencies, specified repository and provides metadata,
    specified minimum Perl version

0.313 Sat 4 Dec 2010 12:55:18 GMT

  * add more tests

0.312 Fri 3 Dec 2010 12:06:53 GMT

  * add commands() and files() routines

0.311 Fri 3 Dec 2010 05:30:16 GMT

  * add more docs
  * fix subcommand() to support old genre

0.31 Thu 25 Nov 2010 03:58:47 PM CST

  * doc update

0.3 Wed 24 Nov 2010 01:32:10 AM CST

  * new cascading attribute for subcommands. thanks to @shelling++.

0.2 Tue 23 Nov 2010 10:51:32 PM CST

  * New Feature: Cascading Subcommand
  * Deprecated Feature: Long Option Subcommand (still work)
  * Abundant DOCs

0.103 - Mon 25 Oct 2010 12:08:58 AM CST

  * Release

0.102 - Wed 20 Oct 2010 02:02:56 AM JST

  * Re-upload

0.10 - 19 Oct 2010

  * Update document

0.09 - 17 Oct 2010

  * Fix the command pattern used in help.

* 0.08 - 25 Feb 2009

  * Allow commands to be provided as inner packages

* 0.07 - 24 Nov 2006

  * Don't use Carp without actually loading it.

* 0.06 - 30 Jun 2006

  * Really fix signatures.

* 0.05 - 5 Jun 2006

  * Fix signatures.

* 0.04 - 11 Apr 2006

  * Delay loading Pod::Simple::Text.

* 0.03 - 27 Dec 2005

  * Forgot to bundle Command::Help.

* 0.02 - 25 Dec 2005

  * Factored out Help system from SVK
  * Factored out more command base stuff from SVK
  (Thanks to Alex Vandiver)


t/lib/MyApp.pm
package MyApp;
use strict;
use warnings;
use base qw(App::CLI App::CLI::Command);

use constant alias => ( te => 'test' );

use constant global_options => (
    'help'       => 'help',
    'username=s' => 'username',
    'force'      => 'force'
);

1;


t/lib/CLITest.pm
package CLITest;
use base 'Exporter';
use strict;
use warnings;
our @EXPORT = qw(cliack clicheck);

our @STACK;

sub cliack {
    push @STACK, [ ( caller(0) )[0], ( caller(1) )[3], @_ ];
}

sub clicheck {
    pop @STACK;
}

1;


t/lib/MyApp/Help.pm
package MyApp::Help;
use base 'App::CLI::Command';

use strict;
use warnings;

sub run {
    my ( $self, @args ) = @_;
}

1;


Makefile.PL
#!/usr/bin/env perl

use strict;
use warnings;

use ExtUtils::MakeMaker qw(6.64);

WriteMakefile(
    NAME          => "App::CLI",
    VERSION_FROM  => 'lib/App/CLI.pm',
    ABSTRACT_FROM => "lib/App/CLI.pm",
    AUTHOR        => [
        'Chia-liang Kao <clkao@clkao.org>',
        'Yo-An Lin      <cornelius.howl@gmail.com>',
        'Shelling       <navyblueshellingford@gmail.com>',
        'Paul Cochrane  <ptc@cpan.org>',
    ],
    LICENSE   => "perl_5",
    PREREQ_PM => {
        'base'                     => 0,
        'constant'                 => 0,
        'strict'                   => 0,
        'warnings'                 => 0,
        'Capture::Tiny'            => 0,
        'Carp'                     => 0,
        'Class::Load'              => 0,
        'File::Basename'           => 0,
        'File::Find'               => 0,
        'Getopt::Long'             => '2.35',
        'Locale::Maketext::Simple' => 0,
        'Pod::Simple::Text'        => 0,
        'Scalar::Util'             => 0,
    },
    MIN_PERL_VERSION => 5.006,
    TEST_REQUIRES    => {
        'lib'            => 0,
        'Class::Load'    => 0,
        'Test::Kwalitee' => 0,
        'Test::More'     => 0,
        'Test::Pod'      => 0,
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        'resources' => {
            'bugtracker' => {
                'web' => 'https://github.com/paultcochrane/app-cli/issues',
            },
            'repository' => {
                'type' => 'git',
                'url'  => 'git@github.com:paultcochrane/app-cli.git',
                'web'  => 'https://github.com/paultcochrane/app-cli',
            },
        },
        x_contributors => [
            'José Joaquín Atria <jjatria@gmail.com>',
            'sunnavy            <sunnavy@gmail.com>',
            'Ildar Shaimordanov <ildar.shaimordanov@gmail.com>',
        ],
    },
);

# vim: expandtab shiftwidth=4


lib/App/CLI/Helper.pm
package App::CLI::Helper;

use strict;
use warnings;

use File::Basename qw( basename );

sub import {
    no strict 'refs';
    my $caller = caller;
    for (qw(commands files prog_name)) {
        *{ $caller . "::$_" } = *$_;
    }
}

=head3 commands()

List the application commands.

=cut

sub commands {
    my ( $class, $include_alias ) = @_;
    my $dir = ref($class) || $class;

    $dir =~ s{::}{/}g;
    $dir = $INC{ $dir . '.pm' };
    $dir =~ s/\.pm$//;

    my @cmds = map { ($_) = m{^\Q$dir\E/(.*)\.pm}; lc($_) } $class->files;

    if ( $include_alias and ref $class and $class->can('alias') ) {
        my %aliases = $class->alias;
        push @cmds, $_ foreach keys %aliases;
    }
    my @sorted_cmds = sort @cmds;

    return @sorted_cmds;
}

=head3 prog_name()

The name of the program running your application. This will default to
C<basename $0>, but can be overridden from within your application.

=cut

{
    my $default;

    sub prog_name {
        my $self = shift;

        $default = basename $0 unless $default;
        return $default unless ref $self;

        return $self->{prog_name} if defined $self->{prog_name};

        $self->{prog_name} = basename $0;
        return $self->{prog_name};
    }
}

=head3 files()

Return module files of subcommands of first level

=cut

sub files {
    my $class = shift;
    $class = ref($class) if ref($class);
    $class =~ s{::}{/}g;
    my $dir = $INC{ $class . '.pm' };
    $dir =~ s/\.pm$//;
    my @sorted_files = sort glob("$dir/*.pm");

    return @sorted_files;
}

1;


lib/App/CLI/Command.pm
package App::CLI::Command;
use strict;
use warnings;
use Locale::Maketext::Simple;
use Carp ();
use App::CLI::Helper;
use Class::Load qw( load_class );
use Scalar::Util qw( weaken );

=head1 NAME

App::CLI::Command - Base class for App::CLI commands

=head1 SYNOPSIS

    package MyApp::List;
    use base qw(App::CLI::Command);

    use constant options => (
        'verbose'   => 'verbose',
        'n|name=s'  => 'name',
    );

    sub run {
        my ( $self, $arg ) = @_;

        print "verbose" if $self->{verbose};

        my $name = $self->{name}; # get arg following long option --name

        # anything you want this command do
    }

    # See App::CLI for information of how to invoke (sub)command.

=head1 DESCRIPTION


=cut

use constant subcommands => ();
use constant options     => ();

sub new {
    my $class = shift;
    bless {@_}, $class;
}

sub command_options {
    ( ( map { $_ => $_ } $_[0]->subcommands ), $_[0]->options );
}

sub run_command {
    my $self = shift;
    $self->run(@_);
}

sub run {
    my $class = shift;
    Carp::croak ref($class) . " does not implement mandatory method 'run'\n";
}

=head3 subcommand()

    return old genre subcommand of $self;

=cut

sub subcommand {
    my $self = shift;
    my @cmd  = $self->subcommands;
    @cmd = values %{ { $self->options } } if @cmd && $cmd[0] eq '*';
    my $subcmd = undef;
    for ( grep { $self->{$_} } @cmd ) {
        no strict 'refs';  ## no critic
        if ( exists ${ ref($self) . '::' }{ $_ . '::' } ) {
            my %data = %{$self};
            $subcmd = bless( {%data}, ( ref($self) . "::$_" ) );
            last;
        }
    }
    $subcmd ? $subcmd : $self;
}

=head3 cascading()

Return instance of cascading subcommand invoked if it was listed in your
constant subcommands.

=cut

sub cascading {
    my $self = shift;
    if ( my $subcmd = $self->cascadable ) {
        shift @ARGV;
        my %data = %{$self};
        return bless {%data}, $subcmd;
    }
    else {
        die $self->error_cmd( $ARGV[0] );
    }
}

=head3 cascadable()

Return package name of subcommand if the subcommand invoked is in your
constant subcommands, otherwise, return C<undef>.

=cut

sub cascadable {
    my $self = shift;
    my $class = ref $self || $self;
    for ( $self->subcommands ) {
        no strict 'refs';  ## no critic
        my $package_name = $class . '::' . $_;
        load_class $package_name;
        if ( $ARGV[0]
            && (ucfirst( $ARGV[0] ) eq $_)
            && exists ${ $class . '::' }{ $_ . '::' } )
        {
            return $package_name;
        }
    }
    return undef;
}

=head3 app

Return the object referring to the current app.

=cut

sub app {
    my $self = shift;

    if (@_) {
        $self->{app} = shift;
        weaken( $self->{app} );
    }

    return $self->{app};
}

=head3 brief_usage ($file)

Display a one-line brief usage of the command object.  Optionally, a file
could be given to extract the usage from the POD.

=cut

sub brief_usage {
    my ( $self, $file ) = @_;
    open my ($podfh), '<', ( $file || $self->filename ) or return;
    local $/ = undef;
    my $buf  = <$podfh>;
    my $base = ref $self->app;
    my $indent = "    ";
    if ( $buf =~ /^=head1\s+NAME\s*\Q$base\E::(\w+)( - .+)$/m ) {
        print $indent, loc( lc($1) . $2 ), "\n";
    }
    else {
        my $cmd = $file || $self->filename;
        $cmd =~ s/^(?:.*)\/(.*?).pm$/$1/;
        print $indent, lc($cmd), " - ", loc("undocumented") . "\n";
    }
    close $podfh;
}

=head3 usage ($want_detail)

Display usage.  If C<$want_detail> is true, the C<DESCRIPTION>
section is displayed as well.

=cut

sub usage {
    my ( $self, $want_detail ) = @_;
    my $fname = $self->filename;
    my ($cmd) = $fname =~ m{\W(\w+)\.pm$};
    require Pod::Simple::Text;
    my $parser = Pod::Simple::Text->new;
    my $buf;
    $parser->output_string( \$buf );
    $parser->parse_file($fname);

    my $base = ref $self->app;
    $buf =~ s/\Q$base\E::(\w+)/\l$1/g;
    $buf =~ s/^AUTHORS.*//sm;
    $buf =~ s/^DESCRIPTION.*//sm unless $want_detail;
    print $self->loc_text($buf);
}

=head3 loc_text $text

Localizes the body of (formatted) text in C<$text> and returns the
localized version.

=cut

sub loc_text {
    my $self = shift;
    my $buf  = shift;

    my $out = "";
    foreach my $line ( split( /\n\n+/, $buf, -1 ) ) {
        if ( my @lines = $line =~ /^( {4}\s+.+\s*)$/mg ) {
            foreach my $chunk (@lines) {
                $chunk =~ /^(\s*)(.+?)( *)(: .+?)?(\s*)$/ or next;
                my $spaces = $3;
                my $loc = $1 . loc( $2 . ( $4 || '' ) ) . $5;
                $loc =~ s/: /$spaces: / if $spaces;
                $out .= $loc . "\n";
            }
            $out .= "\n";
        }
        elsif ( $line =~ /^(\s+)(\w+ - .*)$/ ) {
            $out .= $1 . loc($2) . "\n\n";
        }
        elsif ( length $line ) {
            $out .= loc($line) . "\n\n";
        }
    }
    return $out;
}

=head3 filename

Return the filename for the command module.

=cut

sub filename {
    my $self  = shift;
    my $fname = ref($self);
    $fname =~ s{::[a-z]+$}{};    # subcommand
    $fname =~ s{::}{/}g;
    return $INC{"$fname.pm"};
}

=head1 SEE ALSO

L<App::CLI>, L<Getopt::Long>

=head1 AUTHORS

Chia-liang Kao E<lt>clkao@clkao.orgE<gt>

Cornelius Lin  E<lt>cornelius.howl@gmail.comE<gt>

Shelling       E<lt>navyblueshellingford@gmail.comE<gt>

Paul Cochrane  E<lt>paul@liekut.deE<gt> (current maintainer)

=head1 COPYRIGHT

Copyright 2005-2006 by Chia-liang Kao E<lt>clkao@clkao.orgE<gt>.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut

1;


lib/App/CLI/Command/Help.pm
package App::CLI::Command::Help;
use strict;
use warnings;
use base qw/App::CLI::Command/;
use File::Find qw(find);
use Locale::Maketext::Simple;
use Pod::Simple::Text;

=head1 NAME

App::CLI::Command::Help

=head1 SYNOPSIS

    package MyApp::Help;
    use base qw(App::CLI::Command::Help);

    sub run {
        my $self = shift;
        # preprocess
        $self->SUPER::run(@_);  # App::CLI::Command::Help would output POD of each command
    }

=head1 DESCRIPTION

Your command class should be capitalized.

To add a help message, just add POD in the command class:

    package YourApp::Command::Foo;


    =head1 NAME

    YourApp::Command::Foo - execute foo

    =head1 DESCRIPTION

    blah blah

    =head1 USAGE

    ....

    =cut

=cut

sub run {
    my $self   = shift;
    my @topics = @_;

    push @topics, 'commands' unless (@topics);

    foreach my $topic (@topics) {
        if ( $topic eq 'commands' ) {
            $self->brief_usage($_) for $self->app->files;
        }
        elsif ( my $cmd = eval { $self->app->get_cmd($topic) } ) {
            $cmd->usage(1);
        }
        elsif ( my $file = $self->_find_topic($topic) ) {
            open my $fh, '<:encoding(UTF-8)', $file or die $!;
            require Pod::Simple::Text;
            my $parser = Pod::Simple::Text->new;
            my $buf;
            $parser->output_string( \$buf );
            $parser->parse_file($fh);

            $buf =~ s/^NAME\s+(.*?)::Help::\S+ - (.+)\s+DESCRIPTION/    $2:/;
            print $self->loc_text($buf);
        }
        else {
            die loc( "Cannot find help topic '%1'.\n", $topic );
        }
    }
    return;
}

sub help_base {
    my $self = shift;
    return ref( $self->app ) . "::Help";
}

my ( $inc, @prefix );

sub _find_topic {
    my ( $self, $topic ) = @_;

    if ( !$inc ) {
        my $pkg = __PACKAGE__;
        $pkg =~ s{::}{/};
        $inc = substr( __FILE__, 0, -length("$pkg.pm") );

        my $base = $self->help_base;
        @prefix = ( loc($base) );
        $prefix[0] =~ s{::}{/}g;
        $base =~ s{::}{/}g;
        push @prefix, $base if $prefix[0] ne $base;
    }

    foreach my $dir ( $inc, @INC ) {
        foreach my $prefix (@prefix) {
            foreach my $basename ( ucfirst( lc($topic) ), uc($topic) ) {
                foreach my $ext ( 'pod', 'pm' ) {
                    my $file = "$dir/$prefix/$basename.$ext";
                    return $file if -f $file;
                }
            }
        }
    }

    return;
}

1;


lib/App/CLI/Command/Version.pm
package App::CLI::Command::Version;

use strict;
use warnings;

use base qw/App::CLI::Command/;

=head1 NAME

App::CLI::Command::Version - Print a preformatted version string

=head1 SYNOPSIS

    package MyApp;
    use base qw(App::CLI App::CLI::Command);

    use constant alias => (
        '--version' => '+App::CLI::Command::Version',
          'version' => '+App::CLI::Command::Version',
        # Other aliases
    );

    # Your app now supports a default version command and option

=head1 DESCRIPTION

This is package provides a default C<version> command modelled after
that of L<App::Cmd>. You can modify the default message by subclassing
this command and overriding its C<run> method, or by modifying it with
eg. L<Class::Method::Modifiers>.

=cut

sub run {
    my ($self) = shift;
    no strict 'refs';
    print sprintf "%s (%s) version %s (%s)\n",
      $self->app->prog_name, ref $self->app, $self->app->VERSION, $0;
}

1;


lib/App/CLI/Command/Commands.pm
package App::CLI::Command::Commands;

use strict;
use warnings;

use base qw/App::CLI::Command/;

=head1 NAME

App::CLI::Command::Commands - Print a list of commands for your app

=head1 SYNOPSIS

    package MyApp;
    use base qw(App::CLI App::CLI::Command);

    # Make your app get a list of commands
    use constant alias => (
        commands => '+App::CLI::Command::Commands',
    );

    1;

=head1 DESCRIPTION

Print a list of commands registered for your application;

=cut

sub run {
    my ($self) = shift;

    my ($longest) = sort { length($b) cmp length($a) } $self->app->commands;
    $longest = length $longest;

    foreach ( $self->app->commands ) {
        my $cmd        = $self->app->get_cmd($_);
        my @components = split /::/, ref $cmd;
        my $name       = lc pop @components;
        printf "    %${longest}s\n", $name;
    }
}

1;


lib/App/CLI.pm
package App::CLI;

use strict;
use warnings;
use 5.006;
use Class::Load qw( load_class );

our $VERSION = '0.52';

=encoding utf8

=head1 NAME

App::CLI - Dispatcher module for command line interface programs

=head1 SYNOPSIS

    package MyApp;
    use base 'App::CLI';        # the DISPATCHER of your App
                                # it's not necessary to put the dispatcher
                                # on the top level of your App

    package main;

    MyApp->dispatch;            # call the dispatcher where you want


    package MyApp::List;
    use base qw(App::CLI::Command); # any (SUB)COMMAND of your App

    use constant options => (
        "h|help"   => "help",
        "verbose"  => "verbose",
        'n|name=s'  => 'name',
    );

    use constant subcommands => qw(User Nickname type); # if you want subcommands
                                                        # automatically dispatch to subcommands
                                                        # when invoke $ myapp list [user|nickname|--type]
                                                        # note 'type' is not capitalized
                                                        # it is a deprecated subcommand

    sub run {
        my ($self, @args) = @_;

        print "verbose" if $self->{verbose};
        my $name = $self->{name}; # get arg following long option --name

        if ($self->{help}) {
            # if $ myapp list --help or $ myapp list -h
            # only output PODs
        } else {
            # do something when invoking $ myapp list
            # without subcommand and --help
        }
    }


    package MyApp::List::User;
    use base qw(App::CLI::Command);
    use constant options => (
        "h|help"  =>  "help",
    );

    sub run {
        my ($self,@args) = @_;
        # code for listing user
    }


    pakcage MyApp::List::Nickname;
    use base qw(App::CLI::Command);
    use constant options => (
        "sort=s"  =>  "sort",
    );

    sub run {
        my ($self,@args) = @_;
        # code for listing nickname
    }


    package MyApp::List::type;   # old genre of subcommand could not cascade infinitely
    use base qw(MyApp::List);    # should inherit its parent's command

    sub run {
        my ($self, @args);
        # run to here when invoking $ myapp list --type
    }


    package MyApp::Help;
    use base 'App::CLI::Command::Help';

    use constant options => (
        'verbose' => 'verbose',
    );

    sub run {
        my ($self, @arg) = @_;
        # do something
        $self->SUPER(@_); # App::CLI::Command::Help would output POD of each command
    }

=head1 DESCRIPTION

C<App::CLI> dispatches CLI (command line interface) based commands
into command classes.  It also supports subcommand and per-command
options.

=head2 Methods

=cut

use App::CLI::Helper;
use Getopt::Long ();

use constant alias          => ();
use constant global_options => ();
use constant options        => ();

sub new {
    my ( $class, @args ) = @_;
    my $self = bless {@args}, $class;
    $self->{'app_argv'} = undef;

    return $self;
}

sub app_argv {
    my $self = shift;

    if (@_) {
        $self->{'app_argv'} = shift;
    }

    return $self->{'app_argv'};
}

sub prepare {
    my $self = shift;
    my $data = {};

    $self->get_opt(
        [qw(no_ignore_case bundling pass_through)],
        opt_map( $data, $self->global_options )
    );

    my $command_name = shift @ARGV;
    my $cmd = $self->get_cmd( $command_name, @_, $data );

    while ( $cmd->cascadable ) {
        $cmd = $cmd->cascading;
    }

    $self->get_opt( [qw(no_ignore_case bundling)],
        opt_map( $cmd, $cmd->command_options ) );

    $cmd = $cmd->subcommand;

    return $cmd;
}

=head3 get_opt([@config], %opt_map)

Give options map, processed by L<Getopt::Long::Parser>.

=cut

sub get_opt {
    my ( $self, $config, @app_options ) = @_;
    my $p = Getopt::Long::Parser->new;
    $p->configure(@$config);
    my $err = '';
    local $SIG{__WARN__} = sub {
        my $msg = shift;
        $err .= "$msg";
    };
    my @current_argv = @ARGV;
    $self->app_argv( \@current_argv );
    die $self->error_opt($err) unless $p->getoptions(@app_options);
}

sub opt_map {
    my ( $self, %opt ) = @_;
    return
      map { $_ => ref( $opt{$_} ) ? $opt{$_} : \$self->{ $opt{$_} } } keys %opt;
}

=head3 dispatch(@args)

Interface of dispatcher

=cut

sub dispatch {
    my ( $self, @args ) = @_;
    $self = $self->new unless ref $self;

    $self->app($self) if $self->can('app');

    my $cmd = $self->prepare(@args);
    $cmd->run_command(@ARGV);
}

=head3 cmd_map($cmd)

Find the name of the package implementing the requested command.

The command is first searched for in C<alias>. If the alias exists and points
to a package name starting with the C<+> sign, then that package name (minus
the C<+> sign) is returned. This makes it possible to map commands to arbitrary
packages.

Otherwise, the package is searched for in the result of calling C<commands>,
and a package name is constructed by upper-casing the first character of the
command name, and appending it to the package name of the app itself.

If both of these fail, and the command does not map to any package name,
C<undef> is returned instead.

=cut

sub cmd_map {
    my ( $self, $cmd ) = @_;

    my %alias = $self->alias;

    if ( exists $alias{$cmd} ) {
        $cmd = $alias{$cmd};

        # Alias points to package name, return immediately
        return $cmd if $cmd =~ s/^\+//;
    }

    ($cmd) = grep { $_ eq $cmd } $self->commands;

    # No such command
    return unless $cmd;

    my $base = ref $self;
    return join '::', $base, ucfirst $cmd;
}

sub error_cmd {
    my ( $self, $pkg ) = @_;

    my $cmd;
    if ( defined($pkg) ) {
        $cmd = ref($pkg) || $pkg;
    }
    elsif ( ${ $self->app_argv }[0] ) {
        $cmd = ${ $self->app_argv }[0];
    }
    else {
        $cmd = '<empty command>';
    }

    return "Command '$cmd' not recognized, try $0 --help.\n";
}

sub error_opt { $_[1] }

=head3 get_cmd($cmd, @arg)

Return subcommand of first level via C<$ARGV[0]>.

=cut

sub get_cmd {
    my ( $self, $cmd, $data ) = @_;
    die $self->error_cmd($cmd) unless $cmd && $cmd eq lc($cmd);

    my $base = ref $self;
    my $pkg  = $self->cmd_map($cmd);

    die $self->error_cmd($cmd) unless $pkg;

    load_class $pkg;

    die $self->error_cmd($cmd) unless $pkg->can('run');

    my @arg = defined $data ? %$data : ();
    $cmd = $pkg->new(@arg);
    $cmd->app($self);
    return $cmd;
}

=head1 SEE ALSO

=over 4

=item *

L<App::CLI::Command>

=item *

L<Getopt::Long>

=back

=head1 AUTHORS

=over 4

=item *

Chia-liang Kao E<lt>clkao@clkao.orgE<gt>

=item *

Alex Vandiver  E<lt>alexmv@bestpractical.comE<gt>

=item *

Yo-An Lin      E<lt>cornelius.howl@gmail.comE<gt>

=item *

Shelling       E<lt>navyblueshellingford@gmail.comE<gt>

=item *

Paul Cochrane  E<lt>paul@liekut.deE<gt> (current maintainer)

=back

=head1 CONTRIBUTORS

The following people have contributed patches to the project:

=over 4

=item *

José Joaquín Atria E<lt>jjatria@gmail.comE<gt>

=item *

sunnavy            E<lt>sunnavy@gmail.comE<gt>

=item *


Ildar Shaimordanov E<lt>ildar.shaimordanov@gmail.comE<gt>

=back

=head1 COPYRIGHT

Copyright 2005-2010 by Chia-liang Kao E<lt>clkao@clkao.orgE<gt>.
Copyright 2010 by Yo-An Lin E<lt>cornelius.howl@gmail.comE<gt>
and Shelling E<lt>navyblueshellingford@gmail.comE<gt>.
Copyright 2017-2018 by Paul Cochrane E<lt>paul@liekut.deE<gt>

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

See L<http://www.perl.com/perl/misc/Artistic.html>

=cut

1;
